src/alloc.c

```diff
@@ -832,9 +832,22 @@ JL_DLLEXPORT jl_datatype_t *jl_new_uninitialized_datatype(size_t nfields, int8_t
 unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t) {
     if (!is_vecelement_type(t))
         return 0;
-    if (nfields>16 || (1<<nfields & 0x1157C) == 0)
-        // Number of fields is not 2, 3, 4, 5, 6, 8, 10, 12, or 16.
+    // LLVM 3.7 and 3.8 either crash or generate wrong code for many
+    // SIMD vector sizes N. It seems the rule is that N can have at
+    // most 2 non-zero bits. (This is true at least for N<=100.) See
+    // also <https://llvm.org/bugs/show_bug.cgi?id=27708>.
+    if (nfields <= 32 && (1U << (unsigned)(nfields-1) & 0x808b8bbfU) == 0)
```

Comment:
I'd prefer to not treat `nfields <= 32` as a special case since that introduces another boundary case/path for testing.  The following sequence efficiently detects if `nfields` has two or fewer bits:
```
size_t mask = nfields;
mask &= mask-1;    // Clear least-significant 1 if present
mask &= mask-1;    // Clear another 1
if (mask)
    return 0;      // nfields has more than two 1s. 
```
The clear-least-significant-1 hack is a good one to know in general.  Here's a [ref](https://graphics.stanford.edu/~seander/bithacks.html)  that has it (dates back to [1960](https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetKernighan)!) and other bit hacks.